core/arch/arm/kernel/delay.c

```diff
@@ -27,15 +27,30 @@
  * POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <kernel/boot.h>
 #include <kernel/delay.h>
+#include <kernel/misc.h>
 
+#ifdef CFG_CORE_HAS_GENERIC_TIMER
 void udelay(uint32_t us)
 {
 	uint64_t target = timeout_init_us(us);
 
 	while (!timeout_elapsed(target))
 		;
 }
+#else
+
+void udelay(uint32_t us)
+{
+	uint64_t cycles = 0;
+
+	cycles = (uint64_t)us * ((uint64_t)plat_get_freq() / 1000000ULL);
+	assert(cycles < (unsigned long)-1);
```

Comment:
What is the purpose of this assert? When do we care if there's an overflow or not?

Reply:
Since the `udelay()` is implemented in assembly, it uses 32bits registers and I did not wanted to handle 32/64bits in assembly code, that's why I added this assert here.

Reply:
Using an assert means that you don't care about this in production builds, only during development.
Perhaps it would be better to just set a max value instead in case of overflow.

Reply:
Yes, sorry, indeed, using a maximum value would actually be better.

Reply:
Actually, what can be done is simply to loop on the `wait_cycles` function by calling it with `MIN(cycles, UINT32T_MAX)` and this will handle the 64 bits values.

Reply:
Does it make sense to spin for such a long time? If we're close to UINT32_MAX clearly something is wrong.

Reply:
Actually, the `uint32_t` is meant to hold the number of cycles which might be way greater than the number of usec.

Reply:
OK, got it. Looping as you suggest makes sense.